pandas/core/dtypes/concat.py

```diff
@@ -45,6 +46,8 @@ def get_dtype_kinds(l):
             # if to_concat contains different tz,
             # the result must be object dtype
             typ = str(arr.dtype)
+        elif is_range(arr):
```

Comment:
> We don't have a special 'type' for this index so is_int64_dtype would not work here.

I had missed this, and hence maybe our misunderstanding: ``pd.RangeIndex(3).dtype`` is ``dtype('int64')``... which makes sense, but is not what we want ``get_dtype_kinds`` to consider.